bindings/rust/bench/src/harness.rs

```diff
@@ -1,163 +1,356 @@
+pub trait TlsConnection: Sized {
+    /// Library-specific config struct
+    type Config;
+
+    /// Name of the connection type
+    fn name() -> String;
+
+    /// Default connection (client or server)
+    fn default(mode: Mode) -> Result<Self, Box<dyn Error>> {
+        Self::new(
+            mode,
+            CryptoConfig::default(),
+            HandshakeType::default(),
+            ConnectedBuffer::default(),
+        )
     }
 
+    /// Make a config with given parameters
+    fn make_config(
+        mode: Mode,
+        crypto_config: CryptoConfig,
+        handshake_type: HandshakeType,
+    ) -> Result<Self::Config, Box<dyn Error>>;
+
+    /// Make connection from existing config and buffer
+    fn new_from_config(
+        config: &Self::Config,
+        connected_buffer: ConnectedBuffer,
+    ) -> Result<Self, Box<dyn Error>>;
+
     /// Initialize buffers, configs, and connections (pre-handshake)
     fn new(
+        mode: Mode,
         crypto_config: CryptoConfig,
         handshake_type: HandshakeType,
         buffer: ConnectedBuffer,
-    ) -> Result<Self, Box<dyn Error>>;
+    ) -> Result<Self, Box<dyn Error>> {
+        Self::new_from_config(
+            &Self::make_config(mode, crypto_config, handshake_type)?,
+            buffer,
+        )
+    }
 
-    /// Run handshake on initialized connections
-    /// Returns error if handshake has already completed
+    /// Run one handshake step: receive msgs from other connection, process, and send new msgs
     fn handshake(&mut self) -> Result<(), Box<dyn Error>>;
 
-    /// Checks if handshake is finished for both client and server
     fn handshake_completed(&self) -> bool;
 
-    /// Get negotiated cipher suite
     fn get_negotiated_cipher_suite(&self) -> CipherSuite;
 
-    /// Get whether or negotiated version is TLS1.3
     fn negotiated_tls13(&self) -> bool;
 
-    /// Send application data from connection in harness pair
-    fn send(&mut self, sender: Mode, data: &[u8]) -> Result<(), Box<dyn Error>>;
+    /// Send application data to ConnectedBuffer
+    fn send(&mut self, data: &[u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Read application data from ConnectedBuffer
+    fn recv(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+
+    /// Shrink buffers owned by the connection
+    fn shrink_connection_buffers(&mut self);
 
-    /// Receive application data sent to connection in harness pair
-    fn recv(&mut self, receiver: Mode, data: &mut [u8]) -> Result<(), Box<dyn Error>>;
+    /// Clear and shrink buffers used for IO with another connection
+    fn shrink_connected_buffer(&mut self);
 
-    /// Send data from client to server and then from server to client
-    fn round_trip_transfer(&mut self, data: &mut [u8]) -> Result<(), Box<dyn Error>> {
+    /// Get reference to internal connected buffer
+    fn connected_buffer(&self) -> &ConnectedBuffer;
+}
+
+pub struct TlsConnPair<C: TlsConnection, S: TlsConnection> {
```

Comment:
```suggestion
pub struct TlsConnPair<C: TlsConnection> {
```
Client and server are the same type so you can just use one type for them both.

Reply:
They don't have to be, which is a great thing! Especially for memory benching, I played around with having the connections be different types, which was interesting.

Reply:
They are the same type. The type is "impl TlsConnection". So you don't need separate generics for them.

Reply:
If you wanted to make for example a `TlsConnPair` with a s2n-tls client and Rustls server, you'd need both the `C` and the `S` generic, even though both of them are impls of the same `TlsConnection` trait. For this example, it'd be `TlsConnPair<S2NConnection, RustlsConnection>`.

Reply:
Ah I see what you mean, nvm.